Add topic guide for the suggest component (#6395)
mozilla/application-services: docs/topic-guides/kotlin/suggest.md

```kotlin
import mozilla.appservices.remotesettings.RemoteSettingsServer
import mozilla.appservices.suggest.SuggestApiException
import mozilla.appservices.suggest.SuggestIngestionConstraints
import mozilla.appservices.suggest.SuggestStore
import mozilla.appservices.suggest.SuggestStoreBuilder
import mozilla.appservices.suggest.Suggestion
import mozilla.appservices.suggest.SuggestionQuery
```

Create a `SuggestStore` as a singleton. You do this via the `SuggestStoreBuilder`, which returns a `SuggestStore`.  No I/O or network requests are performed during construction, which makes this safe to do at any point in the application startup:

```kotlin
internal val store: SuggestStore = {
    SuggestStoreBuilder()
        .dataPath(context.getDatabasePath(DATABASE_NAME).absolutePath)
        .remoteSettingsServer(remoteSettingsServer)
        .build()
```

* You need to set the `dataPath`, which is the path (the SQLite location) where you store your suggestions.
* The `remoteSettingsServer` is only needed if you want to set the server to anything else but `prod`. If so, you pass a `RemoteSettingsServer` object. 

## Ingesting suggestions

> We have to handle different phases of the ingesting suggestions: The initial one on starting up the application, and then, periodically, in the background.

Reviewer: ```suggestion
Ingesting suggestions happens in two different ways: On startup, and then, periodically, in the background.
```

Here's my attempt to wordsmith this one, I'm open to other wordings as well.  The main things I'm concerned about are that "we" could be a bit confusing since there are multiple teams involved and also "phases" could imply that the application only needs to do one initial `emptyOnly=true` ingestion, but it should do it every startup to handle upgrades.  The reason for this is that a lot of times when we change the DB schema, we also throw away all the suggestions because it's easier than doing a real migration (and maybe we can't migrate the suggestions without data from remote settings).

* [`SuggestIngestionConstraints`](https://mozilla.github.io/application-services/kotlin/kotlin-components-docs/mozilla.appservices.suggest/-suggest-ingestion-constraints/index.html?query=data%20class%20SuggestIngestionConstraints(var%20providers:%20List%3CSuggestionProvider%3E?%20=%20null,%20var%20providerConstraints:%20SuggestionProviderConstraints?%20=%20null,%20var%20emptyOnly:%20Boolean%20=%20false)) is used to control what gets ingested.

> * Use the `providers` field to limit ingestion by provider type (show an example where you're only ingesting a couple providers. `AMP_MOBILE`, `WIKIPEDIA`, and `WEATHER` are a reasonable combination.

Reviewer: the "show an example" part was for you :)  How about updating the examples below to add the parameter: `providers=listOf(SuggestionProvider.AMP_MOBILE, SuggestionProvider.WIKIPEDIA, SuggestionProvider.WEATHER)`

Author: Oh my, this was a copy/paste mistake!!

> * Use the `providerConstraints` field to add additional constraints, currently this is only used for exposing suggestions.

Reviewer: ```suggestion
* Use the `providerConstraints` field to add additional constraints, currently this is only used for exposure suggestions.
```

"exposure suggestions" are a new suggestion type that the SnG team is introducing.  They're going to be used to measure if a user would be exposed to a suggestion type if it was introduced with a given set of keywords.  The user actually won't be exposed to them, it's all handled internally.

### On Start Up
Ingest with `SuggestIngestionConstraints(emptyOnly=true)` shortly after each startup. This ensures we have something in the DB on the first run and also after upgrades where we often will clear the DB to start from scratch.

```kotlin
store.value.ingest(SuggestIngestionConstraints(emptyOnly = true))

> ```

Reviewer: Is it worth it to duplicate the `providers` field from below here?  That could reinforce the idea that they should be the same for both calls.

Author: Added the `providers` field to both Swift and Kotlin here as well 👍

### Periodically

Ingest with `SuggestIngestionConstraints(emptyOnly=false)` on regular schedule (like once a day).

Example:

```kotlin
store.value.ingest(SuggestIngestionConstraints(emptyOnly = false))
```

## Querying Suggestions

> Call `SuggestStore::query` to fetch suggestions for the suggest bar. The `providerConstraints` should be the same value that got passed to `ingest()`.

Reviewer: ```suggestion
Call `SuggestStore::query` to fetch suggestions for the suggest bar. The `providers` parameter should be the same value that got passed to `ingest()`.
```

# Suggest

The API for the `SuggestStore` can be found in the [MozillaComponents Kotlin documentation](https://mozilla.github.io/application-services/kotlin/kotlin-components-docs/mozilla.appservices.suggest/-suggest-store/index.html).

> > The `SuggestStore` is a synchronous, which needs to be wrapped in the asynchronous primitive of the target language you are using it in.

Reviewer: I'd normally want this to be expanded, but I think it's okay for now since we plan to update our async handling in the near future.

## Setting up the store

> You need to import the following primitives to work with the `SuggestStore` (these come from the generated `suggest.swift` file, produced by `uniffi`):

Reviewer: Maybe "you need to import one or more of the following primitives".  In some cases consumers won't need to import them all.

```kotlin
import mozilla.appservices.remotesettings.RemoteSettingsServer
import mozilla.appservices.suggest.SuggestApiException
import mozilla.appservices.suggest.SuggestIngestionConstraints
import mozilla.appservices.suggest.SuggestStore
import mozilla.appservices.suggest.SuggestStoreBuilder
import mozilla.appservices.suggest.Suggestion
import mozilla.appservices.suggest.SuggestionQuery
```

> On start up of your application, you create a `SuggestStore` (as a singleton). You do this via the `SuggestStoreBuilder`, which returns a `SuggestStore`:

Reviewer: ```suggestion
Create a `SuggestStore` as a singleton. You do this via the `SuggestStoreBuilder`, which returns a `SuggestStore`.   No IO or network requests are performed during construction, which makes this safe to do at any point in the application startup:
```

```kotlin

> internal val store: Lazy<SuggestStore> = lazy {

Reviewer: There's no need for the `lazy` on the consumer size because we lazily load the database nowadays (https://github.com/mozilla/application-services/blob/b10c1319e74041b2777e07657fc83dc879f841f4/components/suggest/src/store.rs#L317).

    SuggestStoreBuilder()
        .dataPath(context.getDatabasePath(DATABASE_NAME).absolutePath)
        .remoteSettingsServer(remoteSettingsServer)
        .build()
```

> In addition to the `dataPath`, which is the path (the SQLite location) where you store your suggestions, you need a `remoteSettingsServer`, which is the URL to the instance of `RemoteSettings` you will be using (e.g. `https://remote-settings-dev.allizom.org`)

Reviewer: The `remoteSettingsServer` is only needed if you want to use something other than `prod`.  They pass in a `RemoteSettingsServer` object rather than a raw URL (https://mozilla.github.io/application-services/kotlin/kotlin-components-docs/mozilla.appservices.remotesettings/-remote-settings-server/index.html).

Author: > The remoteSettingsServer is only needed if you want to use something other than prod.

I don't quite understand. I assumed we always pass the object, just with a different Url?

Reviewer: I think it's optional and you could skip the `.remoteSettingsServer(remoteSettingsServer)` call.  In practice though, consumers will probably always pass it so maybe it's not worth mentioning.

## Ingesting suggestions

> We have to handle three different phases of the ingesting suggestions: The initial one on starting up the application, periodically in the background, and aborting an ongoing ingestion.

Reviewer: I personally don't consider aborting a phase, maybe you could leave that out here and just discuss it at the bottom.

### On Start Up
Ingest with `SuggestIngestionConstraints(emptyOnly=true)` shortly after each startup. This ensures we have something in the DB on the first run and also after upgrades where we often will clear the DB to start from scratch.

```kotlin
store.value.ingest(SuggestIngestionConstraints(emptyOnly = true))
```

### Periodically

Ingest with `SuggestIngestionConstraints(emptyOnly=false)` on regular schedule (like once a day).

Example:

```kotlin
store.value.ingest(SuggestIngestionConstraints(emptyOnly = false))
```

> The defintion of `SuggestIngestionConstraints`:

Reviewer: I'd rather this be more example-focused.   How about saying something like:
  - Starting with "SuggestIngestionConstraints" is used to control what gets ingested (link to the API docs)
  - Use the `providers` field to limit ingestion by provider type (show an example where you're only ingesting a couple providers.   `AMP_MOBILE`, `WIKIPEDIA`, and `WEATHER` are a reasonable combination.
  - Use the `providerConstraints` field to add additional constraints, currently this is only used for exposure suggestions.  (I don't have a great example for this field. @0c0w3 do you have any suggestions?)

```kotlin
/**
 * Constraints limit which suggestions to ingest from Remote Settings.
 */
data class SuggestIngestionConstraints(
    var `providers`: List<SuggestionProvider>?,
    var `providerConstraints`: SuggestionProviderConstraints?,
    /**
     * Only run ingestion if the table `suggestions` is empty

     */
    var `emptyOnly`: kotlin.Boolean,
) {
    companion object
}
```

## Querying Suggestions

> Call `SuggestStore::query` to fetch suggestions for the suggest bar:

Reviewer: `providerConstraints` should be the same value that gets passed to `ingest()`
